### PR TITLE
Refactor filter for BigQuery crawler [sc-6267]

### PR DESCRIPTION
**Pull request by elic-eon**
### 🤔 Why?

To use a general dataset filter.

### 🤓 What?

Remove regular expression based filter, and apply general dataset filter.

### 🧪 Tested?

Manually test BigQuery connector successfully.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -28,6 +28,7 @@ repos:
           - id: mypy
             additional_dependencies:
                 - types-requests
+                - types-freezegun
```
Comment (mars-lan) [inline .pre-commit-config.yaml:31]: Order lexographically.

File: metaphor/bigquery/README.md [hunk 2]
```diff
@@ -39,12 +39,23 @@ See [Common Configurations](../common/README.md) for more information on `output
 
 ### Optional Configurations
 
-You can also optionally specify a list of regular expressions to limit the scan to selected datasets. For example,
+See [Filter Configurations](../common/docs/filter.md) for more information on the optional `filter` config.
+
+#### Notes
+
+Make sure to set `project_id` to `database` field in the filter configuration. For example:
```
Comment (mars-lan) [inline metaphor/bigquery/README.md:46]:
```suggestion
Make sure to use BigQuery project ID when setting the `database` field in the filter configuration. For example:
```

File: metaphor/common/filter.py [hunk 3]
```diff
@@ -51,6 +51,42 @@ def normalize_schema_filter(
 
         return DatasetFilter(includes=includes, excludes=excludes)
 
+    def include_table(self, database: str, schema: str, table: str) -> bool:
+        return include_table(database, schema, table, self)
+
+    def include_schema(self, database: str, schema: str) -> bool:
```
Comment (mars-lan) [inline metaphor/common/filter.py:57]: Please add tests for this.

Comment (mars-lan) [reply metaphor/common/filter.py:57]: Also for consistency, let's add a global function `include_schema` and invoke it from here like `include_table`?

Comment (mars-lan) [reply metaphor/common/filter.py:57]: Actually better yet, get rid of the global `include_table` function and move its content to the `include_table` class method?

Comment (elic-eon) [reply metaphor/common/filter.py:57]: Sounds good, will update to class methods.

File: metaphor/bigquery/README.md [hunk 4]
```diff
@@ -39,12 +39,23 @@ See [Common Configurations](../common/README.md) for more information on `output
...
+project_id: <bigquery_project_id>
+filter:
+  includes:
+    <bigquery_project_id>:
+      schema1:
+      schema2:
+  excludes:
+    <bigquery_project_id>:
+      schema1:
+        - table1
```
Comment (mars-lan) [inline metaphor/bigquery/README.md:58]: Actually just realized we could potentially make this a dict instead of list.... :(

Comment (mars-lan) [reply metaphor/bigquery/README.md:58]: Oh wait never mind, we're already using set internal :P I guess leading `-` looks better than trailing `:`

File: tests/common/test_filter.py [hunk 5]
```diff
@@ -73,8 +73,76 @@ def test_include_table_excludes_overrides_include():
...
+
+    assert not filter.include_schema("db2", "schema1")
+
+    # partial exclude
+    assert filter.include_schema("db2", "schema2")
+
+    assert filter.include_schema("db3", "foo")
+
+
+def test_include_schema_excludes_overrides_include():
```
Comment (mars-lan) [inline tests/common/test_filter.py:135]:
```suggestion
def test_include_schema_excludes_overrides_includes():
```

File: tests/common/test_filter.py [hunk 6]
```diff
@@ -73,8 +73,76 @@ def test_include_table_excludes_overrides_include():
...
+    filter = DatasetFilter(
+        includes={
+            "db1": None,
+            "db2": {"schema1": None, "schema2": set(["table1", "table2"])},
+        },
+        excludes=None,
+    )
+
+    assert filter.include_schema("db1", "foo")
+
```
Comment (mars-lan) [inline tests/common/test_filter.py:105]: Drop the empty line to improve readability?

**Review by mars-lan — COMMENTED** [review]
Otherwise LGTM

**Review by mars-lan — APPROVED** [review]

**Pull request merged**